File: ios/LocationSharingApp/LocationSharingApp/Data/Auth/JwtTokenManager.swift
import Foundation
import Security

/**
 * Manages JWT tokens including storage, retrieval, and validation.
 * Uses enhanced secure token manager for improved security.
 * 
 * @deprecated Use SecureTokenManager directly for new implementations
 */
class JwtTokenManager {
    
    // MARK: - Constants
    
    private struct Constants {
        static let accessTokenKey = "access_token"
        static let refreshTokenKey = "refresh_token"
        static let tokenExpiresAtKey = "token_expires_at"
        static let tokenRefreshBufferSeconds: TimeInterval = 5 * 60 // 5 minutes
        static let keychainService = "com.locationsharing.app.auth"
    }
    
    // MARK: - Singleton
    
    static let shared = JwtTokenManager()
    
    // MARK: - Private Properties
    
    private let secureTokenManager = SecureTokenManager.shared
    
    private init() {}
    
    // MARK: - Token Storage
    
    /**
     * Stores authentication tokens securely.
     * Delegates to SecureTokenManager for enhanced security.
     * 
     * - Parameters:
     *   - accessToken: The JWT access token
     *   - refreshToken: The refresh token
     *   - expiresAt: The timestamp when the access token expires
     */
    func storeTokens(accessToken: String, refreshToken: String, expiresAt: Int64) async throws {
        // Extract user ID from token for secure storage
        let userId = extractUserIdFromToken(accessToken) ?? "unknown"
        try await secureTokenManager.storeTokens(
            accessToken: accessToken,
            refreshToken: refreshToken,
            expiresAt: expiresAt,
            userId: userId
        )
    }
    
    /**
     * Retrieves the stored access token.
     * 
     * - Returns: The access token or nil if not found
     */
    func getAccessToken() async -> String? {
        return try? await secureTokenManager.getAccessToken()
    }
    
    /**
     * Retrieves the stored refresh token.
     * 
     * - Returns: The refresh token or nil if not found
     */
    func getRefreshToken() async -> String? {
        return try? await secureTokenManager.getRefreshToken()
    }
    
    /**
     * Gets the token expiration timestamp.
     * 
     * - Returns: The expiration timestamp or 0 if not found
     */
    func getTokenExpiresAt() async -> Int64 {
        return (try? await secureTokenManager.getTokenExpiresAt()) ?? 0
    }
    
    // MARK: - Token Validation
    
    /**
     * Checks if the current access token is valid and not expired.
     * 
     * - Returns: True if the token is valid, false otherwise
     */
    func isTokenValid() async -> Bool {
        return (try? await secureTokenManager.isTokenValid()) ?? false
    }
    
    /**
     * Checks if the token needs to be refreshed.
     * 
     * - Returns: True if the token should be refreshed, false otherwise
     */
    func shouldRefreshToken() async -> Bool {
        return (try? await secureTokenManager.shouldRefreshToken()) ?? false
    }
    
    /**
     * Extracts user ID from JWT token payload.
     * 
     * - Parameter token: The JWT token
     * - Returns: The user ID or nil if extraction fails
     */
    func extractUserIdFromToken(_ token: String) -> String? {
        let parts = token.components(separatedBy: ".")
        guard parts.count == 3 else { return nil }
        
        guard let payloadData = Data(base64URLEncoded: parts[1]),
              let payloadString = String(data: payloadData, encoding: .utf8) else {
            return nil
        }
        
        // Simple extraction - in production, use a proper JWT library
        let regex = try? NSRegularExpression(pattern: "\"sub\":\"([^\"]+)\"")
        let range = NSRange(payloadString.startIndex..., in: payloadString)
        
        if let match = regex?.firstMatch(in: payloadString, range: range),
           let subRange = Range(match.range(at: 1), in: payloadString) {
            return String(payloadString[subRange])
        }
        
        return nil
    }
    
    /**
     * Checks if a JWT token is expired based on its payload.
     * 
     * - Parameter token: The JWT token
     * - Returns: True if the token is expired, false otherwise
     */
    func isTokenExpired(_ token: String) -> Bool {
        let parts = token.components(separatedBy: ".")
        guard parts.count == 3 else { return true }
        
        guard let payloadData = Data(base64URLEncoded: parts[1]),
              let payloadString = String(data: payloadData, encoding: .utf8) else {
            return true
        }
        
        let regex = try? NSRegularExpression(pattern: "\"exp\":(\\d+)")
        let range = NSRange(payloadString.startIndex..., in: payloadString)
        
        if let match = regex?.firstMatch(in: payloadString, range: range),
           let expRange = Range(match.range(at: 1), in: payloadString),
           let exp = Int64(payloadString[expRange]) {
            let expMs = exp * 1000 // Convert to milliseconds
            let currentMs = Int64(Date().timeIntervalSince1970 * 1000)
            return expMs <= currentMs
        }
        
        return true // If no exp claim or parsing fails, consider expired
    }
    
    /**
     * Clears all stored authentication tokens.
     */
    func clearTokens() async throws {
        try await secureTokenManager.clearTokens()
    }
    
    /**
     * Checks if any authentication tokens are stored.
     * 
     * - Returns: True if tokens exist, false otherwise
     */
    func hasStoredTokens() async -> Bool {
        return (try? await secureTokenManager.hasStoredTokens()) ?? false
    }
    
    // MARK: - Private Keychain Methods
    
    private func storeInKeychain(key: String, value: String) throws {
        let data = value.data(using: .utf8)!
        
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.keychainService,
            kSecAttrAccount as String: key,
            kSecValueData as String: data
        ]
        
        // Delete existing item first
        SecItemDelete(query as CFDictionary)
        
        // Add new item
        let status = SecItemAdd(query as CFDictionary, nil)
        
        guard status == errSecSuccess else {
            throw AuthError.unknown("Failed to store token in keychain: \(status)")
        }
    }
    
    private func retrieveFromKeychain(key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.keychainService,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess,
              let data = result as? Data,
              let string = String(data: data, encoding: .utf8) else {
            return nil
        }
        
        return string
    }
    
    private func deleteFromKeychain(key: String) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: Constants.keychainService,
            kSecAttrAccount as String: key
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        
        // It's okay if the item doesn't exist
        guard status == errSecSuccess || status == errSecItemNotFound else {
            throw AuthError.unknown("Failed to delete token from keychain: \(status)")
        }
    }
}

// MARK: - Base64URL Extension

private extension Data {
    init?(base64URLEncoded string: String) {
        var base64 = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        
        // Add padding if needed
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        
        self.init(base64Encoded: base64)
    }
}

File: android/app/src/main/java/com/locationsharing/app/data/auth/JwtTokenManager.kt
package com.locationsharing.app.data.auth

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.Base64
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Manages JWT tokens including storage, retrieval, and validation.
 * Uses encrypted shared preferences for secure token storage.
 */
@Singleton
class JwtTokenManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    
    companion object {
        private const val PREFS_NAME = "auth_prefs"
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
        private const val KEY_TOKEN_EXPIRES_AT = "token_expires_at"
        private const val TOKEN_REFRESH_BUFFER_MS = 5 * 60 * 1000L // 5 minutes
    }
    
    private val masterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()
    
    private val encryptedPrefs: SharedPreferences by lazy {
        EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )
    }
    
    /**
     * Stores authentication tokens securely.
     * 
     * @param accessToken The JWT access token
     * @param refreshToken The refresh token
     * @param expiresAt The timestamp when the access token expires
     */
    suspend fun storeTokens(
        accessToken: String,
        refreshToken: String,
        expiresAt: Long
    ) = withContext(Dispatchers.IO) {
        encryptedPrefs.edit()
            .putString(KEY_ACCESS_TOKEN, accessToken)
            .putString(KEY_REFRESH_TOKEN, refreshToken)
            .putLong(KEY_TOKEN_EXPIRES_AT, expiresAt)
            .apply()
    }
    
    /**
     * Retrieves the stored access token.
     * 
     * @return The access token or null if not found
     */
    suspend fun getAccessToken(): String? = withContext(Dispatchers.IO) {
        encryptedPrefs.getString(KEY_ACCESS_TOKEN, null)
    }
    
    /**
     * Retrieves the stored refresh token.
     * 
     * @return The refresh token or null if not found
     */
    suspend fun getRefreshToken(): String? = withContext(Dispatchers.IO) {
        encryptedPrefs.getString(KEY_REFRESH_TOKEN, null)
    }
    
    /**
     * Gets the token expiration timestamp.
     * 
     * @return The expiration timestamp or 0 if not found
     */
    suspend fun getTokenExpiresAt(): Long = withContext(Dispatchers.IO) {
        encryptedPrefs.getLong(KEY_TOKEN_EXPIRES_AT, 0L)
    }
    
    /**
     * Checks if the current access token is valid and not expired.
     * 
     * @return True if the token is valid, false otherwise
     */
    suspend fun isTokenValid(): Boolean = withContext(Dispatchers.IO) {
        val token = getAccessToken()
        val expiresAt = getTokenExpiresAt()
        val currentTime = System.currentTimeMillis()
        
        token != null && expiresAt > currentTime + TOKEN_REFRESH_BUFFER_MS
    }
    
    /**
     * Checks if the token needs to be refreshed.
     * 
     * @return True if the token should be refreshed, false otherwise
     */
    suspend fun shouldRefreshToken(): Boolean = withContext(Dispatchers.IO) {
        val expiresAt = getTokenExpiresAt()
        val currentTime = System.currentTimeMillis()
        
        expiresAt > 0 && expiresAt <= currentTime + TOKEN_REFRESH_BUFFER_MS
    }
    
    /**
     * Extracts user ID from JWT token payload.
     * 
     * @param token The JWT token
     * @return The user ID or null if extraction fails
     */
    fun extractUserIdFromToken(token: String): String? {
        return try {
            val parts = token.split(".")
            if (parts.size != 3) return null
            
            val payload = String(Base64.getUrlDecoder().decode(parts[1]))
            // Simple extraction - in production, use a proper JWT library
            val userIdRegex = "\"sub\":\"([^\"]+)\"".toRegex()
            userIdRegex.find(payload)?.groupValues?.get(1)
        } catch (e: Exception) {
            null
        }
    }
    
    /**
     * Checks if a JWT token is expired based on its payload.
     * 
     * @param token The JWT token
     * @return True if the token is expired, false otherwise
     */
    fun isTokenExpired(token: String): Boolean {
        return try {
            val parts = token.split(".")
            if (parts.size != 3) return true
            
            val payload = String(Base64.getUrlDecoder().decode(parts[1]))
            val expRegex = "\"exp\":(\\d+)".toRegex()
            val expMatch = expRegex.find(payload)
            
            if (expMatch != null) {
                val exp = expMatch.groupValues[1].toLong() * 1000 // Convert to milliseconds
                exp <= System.currentTimeMillis()
            } else {
                true // If no exp claim, consider expired
            }
        } catch (e: Exception) {
            true // If parsing fails, consider expired
        }
    }
    
    /**
     * Clears all stored authentication tokens.
     */
    suspend fun clearTokens() = withContext(Dispatchers.IO) {
        encryptedPrefs.edit()
            .remove(KEY_ACCESS_TOKEN)
            .remove(KEY_REFRESH_TOKEN)
            .remove(KEY_TOKEN_EXPIRES_AT)
            .apply()
    }
    
    /**
     * Checks if any authentication tokens are stored.
     * 
     * @return True if tokens exist, false otherwise
     */
    suspend fun hasStoredTokens(): Boolean = withContext(Dispatchers.IO) {
        encryptedPrefs.contains(KEY_ACCESS_TOKEN) && 
        encryptedPrefs.contains(KEY_REFRESH_TOKEN)
    }
}